### ISSUE TITLE
Feedback/Request For Simplification In A Collection Runner

### ISSUE THREAD

**Issue by ponmuthu5990** [postmanlabs/postman-collection#1206]
Hi,

This is duplicate of request 80044, As it is asked to post this feedback/request in GitHub public issues.

I Apologize in advance for the lengthy feedback, Couldn't find any simple way to explain this. Also I had put some of the content repeatedly, apologize for that too. Happy reading!!

Let's start with simple collection runner for CURD. For example, A tester going to test CURD APIs with collection,

> **Step 1:** Create a POST request for create data and save id from response to variable. 
> **Step 2:** Use the id from variable for GET request and verify the data created in the test tab and save the response into another variable.
> **Step 3:** Now use the previous response and change values in prescript for PUT API and update the data.
> **Step 4:** Now use the Id variable for GET request and make sure the changes in step 3 got updated, in test tab.
> **Step 5:** Again use id variable in DELETE request for delete data,
> **Step 6:** One last time, use the id variable for GET request, to verify the data deleted and obviously delete all the variables.

 So, A tester can test all the APIs in single button click, But we can do only for positive scenarios. you know like, In all the above 6 steps we are always passing the correct values that APIs are expect. 

But, **As a tester, only way to make sure that the APIs accepting correct value is testing with incorrect values, right ?**
 
For example, In request body, we have variable 'length' which is NotNull and accept only 0 to 100. A tester have to test with -1, 101, null, 'abc', etc. These values should not be accepted and most likely API will throw some errors, say 422 unprocessable entity. 

For a single variable('length'), We have to create 4 requests in a collection, which we expect same response 422 unprocessable entity. Let's Assume a request body with 20 to 30 variable or more than that. This will make around 100 or more negative scenario requests for single API(POST).

If the variable is unique field, like email id or phone number or any other id. You know, after successfully create (Step 1: POST) and verify (Step 2: GET) the data, A tester had to test with same email id or phone number to make sure the duplicate data not getting updated through API. 

Because of these negative scenario requests, CURD API automation/collection becomes very complex. I mean, we have to add the same request multiple times with different data(request body) in collection runner(Below example). Because we can't load JSON file in a single request for iteration. 

let's Assume 100 negative scenario in request(POST) body,

> **Step 1 - Step 100:** Create a 100 POST request and verify 422 error code. (Basically all the 100 request are hitting the same API(POST) and expecting the same response 422, only difference is each request body contain different value -  negative scenario values).
> **Step 101:** Create a POST request for create data and save id from response to variable.
> **Step 102:** Use the id from variable for GET request and verify the data created in the test tab and save the response into another variable.
> **Step 103:** This one is for verifying the unique variable, like email id and phone number to make sure duplicate data is not acceptable.(This is step is an optional one).
> **Step 104 - Step 204:** Create a 100 PUT request and verify the 422 error code. (Basically all the 100 request are hitting the same API(PUT) and expecting the same response 422, only difference is the request body contain different value -  negative scenario values).
> **Step 205:** Now use the previous response from Step -103 and change values in prescript for PUT API and update the data.
> **Step 206:** Now use the Id variable for GET request and make sure the change in Step 205 got updated, in test tab.
> **Step 207:** Again, use id variable in DELETE request for delete data.
> **Step 208:** One last time, use the id variable for GET request, to verify the data deleted and obviously delete all the variable. 

As per the above example, A tester had to create around 210 request in collection for complete testing CURD API in a single button click. But in real time, it goes up to 400 requests in a collection. Also, we have two ways to execute this, right ? either, 

> 1. Create 400 requests in a collection(Most of them are same request- hitting the same endpoint and expecting the same error response code with different data) and execute via collection runner in a single button click.
> 2. Create nested folder/collection for CURD API and execute one by one manually. like, 
>     1. Negative scenario collection for POST,
>     2. Positive scenario collection for POST (Inside this we can create data(POST) then verify data(GET) and then Duplicate scenario(If any) and finally delete data(DELETE) with the help for variable),
>     4. Negative scenario collection for PUT,
>     5. Positive scenario collection for PUT (Inside this we can create data(POST) then update data(PUT) and next verify data(GET) and then Duplicate scenario(If any) and finally delete data(DELETE) with the help for variable),

If we choose the second way, It is looks like okay in structure wise but we have to execute one by one every time and if we generate any custom reports from this(Using Newman), it is getting very complex than it looks. I don't think, we can call it as automation if we execute one by one. Unless if we have JSON file iteration in request level(Currently we have in running a collection), We can do both positive and negative scenario test with less no of request(Below Example) in collection.

> **Step 1:** Create a POST request with 100 or more that in a JSON file for iteration to test negative scenario and verify the 422 error code in test tab. (We can add data as much we want in JSON for iteration, right.)
> **Step 2:** Create a POST request for create data and save id from response to variable.
> **Step 3:**  Use the id from variable for GET request and verify the data created in the test and save the response into another variable.
> **Step 4:** This one is for verifying the unique variable, like email id and phone number to make sure duplicate data is not acceptable.(if we have multiple unique field, we can attach JSON for iteration).
> **Step 5:** Create a PUT request with 100 or more that in a JSON file for iteration to test negative scenario and verify the 422 error code in test tab. (we can add data as much we want in JSON for iteration).
> **Step 6:** Now use the previous response and change values in prescript for PUT API and update the data.
> **Step 7:** Now use the Id variable for GET request and make sure the change in Step 6 got updated in test tab.
> **Step 8:** Again use id variable for DELETE request for delete data.
> **Step 9:** One last time, use the id variable for GET request, to verify the data deleted and obviously delete all the variable. 

Basically, if a tester run this above example via collection runner, each and every request will execute as per the iteration number(depends on JSON file that attached to that particular request) and move on to next request. At the end of this, As tester had 100% sure that CURD APIs never accept invalid values and duplicate values also the valid values are successfully created, updated, deleted and everything verified in a single button click.

**In a simple words, we can have a JSON file attachment for iteration into request level. Like, we have in running a collection.** Please let me know, this whole thing is kinda make sense or not.
